src/App.js

```diff
@@ -1,23 +1,133 @@
-import logo from './logo.svg';
-import './App.css';
+import "./App.css";
+
+import Header from "./components/Header";
+import DisplayAlgo from "./components/DisplayAlgo";
+import DisplayAlgoLine from "./components/DisplayAlgoLine";
+import Footer from "./components/Footer";
+import { useState, useEffect } from "react";
 
 function App() {
+  const numRow = Math.floor(window.innerWidth / 11);
+  const [runningAlgorithm, setRunningAlgorithm] = useState(false);
+  const [rows, setRows] = useState([]);
+  const [arrayAccessCounter, setArrayAccessCounter] = useState(0);
+  const [comparisonCounter, setComparisonCounter] = useState(0);
+  //_____________________Functions________________________________________
+
+  //randomize graph
+  const randomize = () => {
+    const tempArr = [];
+    for (var i = 0; i < numRow; i++) {
+      tempArr.push({
+        height: Math.floor(Math.random() * 500),
+        id: i,
+      });
+    }
+    setRows((rows) => tempArr);
+  };
+
+  //setStates running program
+  const changeRunningAlgorithm = (running) => {
+    if (running) {
+      setRunningAlgorithm(true);
+    } else {
+      setRunningAlgorithm(false);
+    }
+  };
+
+  //____________________Sorters_start___________________________________
+
+  const bubbleSort = () => {
+    let tempArr = [...rows];
+    let tempNum = 0;
+    let tempArrCount = 0;
+
+    changeRunningAlgorithm(true);
+
+    var innerLoop = () => {
+      //inner loop
+      for (let y = 1; y < numRow; y++) {
+        if (tempArr[y].height < tempArr[y - 1].height) {
+          //swap if height is bigger
+          let copy = [...tempArr];
+          [copy[y - 1], copy[y]] = [copy[y], copy[y - 1]];
+          tempArr = copy;
+          tempArrCount++;
+        }
+        tempNum++;
+        tempArrCount = tempArrCount + 2;
+      }
+      //add to States
+      setRows(tempArr);
+      setComparisonCounter(tempNum);
+      setArrayAccessCounter(tempArrCount);
+    };
+
+    //outer loop
+    for (let i = 0; i < numRow; i++) {
+      setTimeout(innerLoop, 50 * i);
+      setTimeout(() => {
+        changeRunningAlgorithm(false);
+      }, 50 * numRow);
+    }
```

Comment:
The second timeout here, which executes the `changeRunningAlgorithm`, is currently being executed `numRows` times - although it _always_ performs the same change. Every single time this happens, and because it's running in a async code (`setTimeout`), this is gonna trigger a **whole component re-render**. This is a BIG performance problem.

As a simple solution, you could simply have one single timeout, outside the loop:

```suggestion
    for (let i = 0; i < numRow; i++) {
      setTimeout(innerLoop, 50 * i);
    }
    
    setTimeout(() => {
        changeRunningAlgorithm(false);
      }, 50 * numRow);
```

But this has a downside: in JavaScript, you cannot ensure postponed code executions will run in order. This means it might happen that the `changeRunningAlgorithm` call gets executed _before_ the last `innerLoop`.

The possible solution would be to have this logic within the `setTimeout` used for the `innerLoop`:

```suggestion
    for (let i = 0; i < numRow; i++) {
      const isLast = i === numRow - 1
      setTimeout(() => {
        innerLoop()
        if (isLast) {
           changeRunningAlgorithm(false);
        }
      }, 50 * i);
    }
```

> In this example, the `isLast` is defined outside the `setTimeout` because of potential [variable scope leaking](https://stackoverflow.com/questions/31285911/why-let-and-var-bindings-behave-differently-using-settimeout-function?lq=1).